### PR TITLE
Improve Microsoft login

### PR DESCRIPTION
**Pull request by Trial97**
This is the start of refactoring the auth stuff.
I want to achieve better error handling and display the message to the user in case something fails(instead fo `Working...`)
Right now this was done:
- [x] Merged NetAction with NetRequest as it was the only class that implemented it
- [x] Replaced AuthRequest with derivates of NetRequest to ensure the same error handling for all HTTP requests 
- [x] Merged AccountTask with AuthFlow as it was the only class that implemented it
- [x] Removed old oauth lib in favor of qt one
- [x] Fix nix build
- [x] fix codeql
- [x] better UI for Login dialog 
- [x] support device code flow

There is more to come so stay tuned.

### CODE REVIEW

**Comment by Trial97** [comment]
Ok, now the basic build works, but I have troubles with codeql and nix stuff so any feedback there is appreciated.

File: launcher/minecraft/Library.cpp
Comment (Trial97) [reply launcher/minecraft/Library.cpp:78]: Did not touch the exact function so leave it uncommented

File: launcher/ui/dialogs/MSALoginDialog.cpp
Comment (TayouVR) [inline launcher/ui/dialogs/MSALoginDialog.cpp]:
```suggestion
           "If your browser hasn't opened, please manually open the below link in your browser:"));
```
maybe even swap below and link, I'm not that great at English grammar either though, its not my first language x3

**Review by TayouVR — APPROVED** [review]
looks to be working as intended, and definitely feels nicer to use than the previous code method. 
Tested it on win10; If nobody else finds any issues with it I'd be happy to get this merged soon :3

**Review by DioEgizio — APPROVED** [review]
this def feels a lot better

**Comment by Trial97** [comment]
I added also the device code flow back

**Comment by TayouVR** [comment]
btw can we customize the page that is shown after completing login? Or is that fixed in the QNetworkAuth thing?
would be neat if we could tint it like our website and add the prism launcher logo in the center or smth
With a text like: "Microsoft Account is now connected with Prism Launcher, you can close this page now."

**Comment by Trial97** [comment]
Yes we can change that.
Do you prefer to have all the design hardcoded in the launcher or have it just an iframe towards prism website, were we can have a path just for that?

**Comment by evan-goode** [comment]
For the Nix build, I think we just need to add `qtnetworkauth` as a build dependency: https://github.com/unmojang/PrismLauncher/commit/73b84f522f1697acec015d422f63ddd542d97de7. Feel free to cherry-pick. Both device-code and auth-code flows work for me on NixOS on both Qt5 and Qt6.

As a user, I would rather not have to decide whether to log in "Using device code" or "Using auth code". Someone might look at that dialog box and say, "Uh, I have neither of those, I have a Microsoft email/password? Which one do I pick?" Is there a good reason to keep the device code login flow? The flows feel almost identical in UX.

**Comment by evan-goode** [comment]
Another note: if the user closes the "How do you want to login?" dialog, the login will not be canceled, instead, the "Using auth code" flow will proceed. (I am using Sway, maybe the window is not closeable on other window managers). I think adding a "Cancel" option to the dialog would fix that, and might be a good idea regardless.

**Comment by Trial97** [comment]
@evan-goode, thank you for the nix suggestion!
Regarding the auth dialog I am considering changing the button text but right now I have no good idea of what the text should be( suggestion accepted).
Regarding the cancel stuff I need to look into it, most probably I will add a cancel button.

File: launcher/ui/pages/global/AccountListPage.cpp
Comment (DioEgizio) [inline launcher/ui/pages/global/AccountListPage.cpp]: shouldn't the default button be "Reccomended"?

Comment (Trial97) [reply launcher/ui/pages/global/AccountListPage.cpp]: I put cancel as the default because the default button is also considered clicked, and if I exit the message box by closing it(not hitting any button) I do not want to proceed with the login.

Comment (DioEgizio) [reply launcher/ui/pages/global/AccountListPage.cpp]: it's still kinda weird to default to cancel, cuz I can see people may just want to just press enter and login, without having to move to recommended

Comment (TayouVR) [reply launcher/ui/pages/global/AccountListPage.cpp]: can't you handle the close button manually? Or does the message box automatically do the reject/cancel action when closed? That would be the expected behaviour to me... 
I can't see anything obvious in the docs, but haven't worked with it so far either

Comment (Trial97) [reply launcher/ui/pages/global/AccountListPage.cpp]: Sorry, I was wrong it looks like you are right. I made the recommended button default.

**Review by DioEgizio — APPROVED** [review]

**Review by TayouVR — APPROVED** [review]

**Comment by Trial97** [comment]
Do I need to wait for another maintainer for this?

**Comment by TayouVR** [comment]
nope, I think we can merge it. We have the 2 approvals thing for a reason ^^

**Pull request merged**
